main/lsp/requests/completion.cc

```diff
@@ -18,43 +18,50 @@ namespace {
 struct RubyKeyword {
     const string keyword;
     const string documentation;
+    const optional<string> snippet;
 
-    RubyKeyword(string keyword, string documentation) : keyword(keyword), documentation(documentation){};
+    RubyKeyword(string keyword, string documentation, optional<string> snippet = nullopt)
+        : keyword(keyword), documentation(documentation), snippet(snippet){};
 };
 
 // Taken from https://docs.ruby-lang.org/en/2.6.0/keywords_rdoc.html
 // We might want to put this somewhere shareable if there are more places that want to use it.
+//
+// VS Code snippet syntax is in general smarter than LSP snippet syntax.
+// Specifically, VS Code will intelligently insert the correct indentation after newlines.
 const RubyKeyword rubyKeywords[] = {
     {"BEGIN", "Runs before any other code in the current file."},
     {"END", "Runs after any other code in the current file."},
     {"__ENCODING__", "The script encoding of the current file."},
     {"__FILE__", "The path to the current file."},
     {"__LINE__", "The line number of this keyword in the current file."},
-    {"alias", "Creates an alias between two methods (and other things)."},
+    {"alias", "Creates an alias between two methods (and other things).", "alias ${1:_new} ${2:_existing}$0"},
     {"and", "Short-circuit Boolean and with lower precedence than &&"},
-    {"begin", "Starts an exception handling block."},
+    {"begin", "Starts an exception handling block.", "begin\n  $0\nend"},
     {"break", "Leaves a block early."},
-    {"case", "Starts a case expression."},
-    {"class", "Creates or opens a class."},
-    {"def", "Defines a method."},
-    {"defined?", "Returns a string describing its argument."},
-    {"do", "Starts a block."},
-    {"else", "The unhandled condition in case, if and unless expressions."},
-    {"elsif", "An alternate condition for an if expression."},
+    {"case", "Starts a case expression.", "case ${1:expr}\nwhen ${2:expr}\n  $0\nelse\nend"},
+    {"class", "Creates or opens a class.", "class ${1:ClassName}\n  $0\nend"},
+    {"def", "Defines a method.", "def ${1:method_name}($2)\n  $0\nend"},
+    {"defined?", "Returns a string describing its argument.", "defined?(${1:Constant})$0"},
+    // TODO(jez) Even better would be to auto-insert a block for methods that we know must take a block
+    {"do", "Starts a block.", "do\n  $0\nend"},
+    {"else", "The unhandled condition in case, if and unless expressions.", "else ${1:expr}$0"},
+    {"elsif", "An alternate condition for an if expression.", "elsif ${1:expr}$0"},
     {"end",
      "The end of a syntax block. Used by classes, modules, methods, exception handling and control expressions."},
     {"ensure", "Starts a section of code that is always run when an exception is raised."},
     {"false", "Boolean false."},
     {"for", "A loop that is similar to using the each method."},
```

Comment:
Is there no good snippet for `for`? (`"for ${1:expr} in ${2:expr}\n  $0\nend"` ?

Reply:
Ah maybe; i just know that we lint against `for` in pay-server so i didn't want to think about what the syntax looked like.